# Гибридная рекомендательная система (User-Based + Item-Based KNN)

## 1. Постановка задачи

Цель работы — разработать, оценить и исследовать гибридную рекомендательную
систему (РС) на базе алгоритма k ближайших соседей (k-NN), объединяющую два
классических подхода коллаборативной фильтрации:

  - User-based Collaborative Filtering (UBCF): рекомендации на основе
    предпочтений похожих пользователей («социальный контекст»).
  - Item-based Collaborative Filtering (IBCF): рекомендации на основе схожести
    объектов («контентно-поведенческий контекст» фильмов, высоко оцененных
    целевым пользователем).

### Задачи:

1.  Провести разведочный анализ данных (EDA) датасета MovieLens 1M.
2.  Реализовать конвейер генерации кандидатов обоими методами с нормализацией
    оценок.
3.  Разработать алгоритм объединения списков: бустинг пересечений (UB \cap IB) и
    ранжирование непересекающихся кандидатов.
4.  Исследовать влияние гиперпараметров (k_{users}, k_{items}, балансировочный
    вес \alpha, штрафы/бонусы).
5.  Провести два углубленных исследования:
      - Качество рекомендаций для полярных групп пользователей (топ 10% активных
        vs 10% неактивных).
      - Стратегия преодоления холодного старта для пользователей без истории
        оценок.

In [10]:
import time
import numpy as np
import pandas as pd
from scipy.sparse import csr_matrix
from sklearn.neighbors import NearestNeighbors
from sklearn.model_selection import train_test_split
import math
import warnings
warnings.filterwarnings('ignore')

# =====================================================================
# ГЛОБАЛЬНЫЕ ГИПЕРПАРАМЕТРЫ МОДЕЛИ
# =====================================================================
GLOBAL_PARAMS = {
    'TEST_SIZE': 0.20,         # 20% данных в контрольную выборку
    'RANDOM_STATE': 42,
    'K_USERS': 30,             # Число соседей-пользователей
    'K_ITEMS': 20,             # Число соседей-фильмов
    'TOP_X': 10,               # Длина итогового списка рекомендаций
    'MIN_RATING_LIKE': 4,      # Порог позитивной оценки для seed-фильмов в Item-Based
    'ALPHA_UB': 0.55,          # Вес User-based составляющей (1 - ALPHA = вес Item-based)
    'INTERSECTION_BONUS': 1.25 # Множитель доверия (бонус 25%) для фильмов из пересечения
}

## 2. Структура и анализ данных (EDA)

Датасет MovieLens 1M содержит 1000209 дискретных рейтингов (от 1 до 5)
от 6040 пользователей к 3706 фильмам. Каждый пользователь оценил не менее 20
фильмов.

In [14]:
def load_and_inspect_data():
    r_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_csv('ratings.dat', sep='::', names=r_cols, engine='python', encoding='latin-1')

    m_cols = ['movie_id', 'title', 'genres']
    movies = pd.read_csv('movies.dat', sep='::', names=m_cols, engine='python', encoding='latin-1')

    u_cols = ['user_id', 'gender', 'age', 'occupation', 'zip']
    users = pd.read_csv('users.dat', sep='::', names=u_cols, engine='python', encoding='latin-1')

    n_users = ratings['user_id'].nunique()
    n_movies = ratings['movie_id'].nunique()
    n_ratings = len(ratings)

    # Расчет разреженности матрицы (Sparsity)
    total_possible = n_users * n_movies
    sparsity = (1.0 - (n_ratings / total_possible)) * 100

    print("=" * 55)
    print("АНАЛИЗ СОДЕРЖАНИЯ ДАННЫХ (MovieLens 1M)")
    print("=" * 55)
    print(f"Уникальных пользователей: {n_users:,}")
    print(f"Уникальных фильмов:       {n_movies:,}")
    print(f"Всего выставлено оценок:  {n_ratings:,}")
    print(f"Разреженность матрицы:    {sparsity:.2f}%")
    print(f"Средний рейтинг по базе:  {ratings['rating'].mean():.2f} / 5.0")
    print("=" * 55)

    return ratings, movies, users

ratings_df, movies_df, users_df = load_and_inspect_data()

АНАЛИЗ СОДЕРЖАНИЯ ДАННЫХ (MovieLens 1M)
Уникальных пользователей: 6,040
Уникальных фильмов:       3,706
Всего выставлено оценок:  1,000,209
Разреженность матрицы:    95.53%
Средний рейтинг по базе:  3.58 / 5.0


## 3. Математический аппарат и подход к реализации

3.1. Метрика расстояния

Используется косинусное расстояние между векторами.

3.2. Компоненты гибридизации

1.  User-Based Score: Для непросмотренных целевым пользователем u
    фильмов прогнозируется оценка как средневзвешенное оценок k ближайших
    соседей N(u):
    \hat{r}_{UB}(u, i) = \frac{\sum_{v \in N(u)} Sim(u, v) \cdot r_{v, i}}{\sum_{v \in N(u)} |Sim(u, v)|}
2.  Item-Based Score (S_{IB}): Берутся фильмы J_u, которым пользователь поставил
    r_{u, j} \ge 4. Для каждого такого фильма находятся ближайшие соседи
    i \notin \text{истории } u. Скор накапливается пропорционально сходству
    Sim(i, j) и оценке r_{u, j}:
    \hat{r}_{IB}(u, i) = \sum_{j \in J_u} Sim(i, j) \cdot \frac{r_{u, j}}{5.0}
3.  Мин-макс нормализация: Поскольку S_{UB} и S_{IB} находятся на разных шкалах,
    они нормируются в отрезок [0, 1] перед смешиванием:
    S'_{UB} = \frac{S_{UB} - \min(S_{UB})}{\max(S_{UB}) - \min(S_{UB}) + \epsilon}, \quad S'_{IB} = \frac{S_{IB} - \min(S_{IB})}{\max(S_{IB}) - \min(S_{IB}) + \epsilon}
4.  Формула пересечения и штрафования/бонусирования:
    S_{Final}(u, i) = \left( \alpha \cdot S'_{UB}(u, i) + (1 - \alpha) \cdot S'_{IB}(u, i) \right) \cdot \beta
    Где:
      - \beta = \text{INTERSECTION\_BONUS} = 1.25, если фильм попал и в UBCF, и
        в IBCF выдачу.
      - \beta = 1.0, если фильм присутствует только в одной из выборок.

In [3]:
def prepare_matrices(ratings, test_size=GLOBAL_PARAMS['TEST_SIZE']):
    train_df, test_df = train_test_split(
        ratings,
        test_size=test_size,
        random_state=GLOBAL_PARAMS['RANDOM_STATE'],
        stratify=None
    )

    # 1. Строки — пользователи, столбцы — фильмы (для User-based)
    ui_matrix = train_df.pivot(index='user_id', columns='movie_id', values='rating').fillna(0)

    # 2. Строки — фильмы, столбцы — пользователи (для Item-based)
    iu_matrix = train_df.pivot(index='movie_id', columns='user_id', values='rating').fillna(0)

    # Разреженные структуры
    ui_sparse = csr_matrix(ui_matrix.values)
    iu_sparse = csr_matrix(iu_matrix.values)

    # Обучение моделей NearestNeighbors
    model_ub = NearestNeighbors(metric='cosine', algorithm='brute', n_neighbors=GLOBAL_PARAMS['K_USERS'], n_jobs=-1)
    model_ub.fit(ui_sparse)

    model_ib = NearestNeighbors(metric='cosine', algorithm='brute', n_neighbors=GLOBAL_PARAMS['K_ITEMS'], n_jobs=-1)
    model_ib.fit(iu_sparse)

    return model_ub, model_ib, ui_matrix, iu_matrix, train_df, test_df

model_ub, model_ib, ui_matrix, iu_matrix, train_df, test_df = prepare_matrices(ratings_df)

In [4]:
def get_hybrid_recommendations(user_id, model_ub, model_ib, ui_matrix, iu_matrix,
                               movies_df, top_x=GLOBAL_PARAMS['TOP_X'],
                               alpha=GLOBAL_PARAMS['ALPHA_UB'],
                               bonus=GLOBAL_PARAMS['INTERSECTION_BONUS'],
                               min_support=2):
    """
    Генерация гибридных рекомендаций (User-Based + Item-Based).

    Параметры:
    - min_support: минимальное число соседей, оценивших фильм, чтобы он попал в UB-кандидаты
    - alpha: вес User-Based составляющей (1 - alpha: вес Item-Based)
    - bonus: коэффициент бустинга пересечения (UB ∩ IB)
    """
    if user_id not in ui_matrix.index:
        return pd.DataFrame()

    user_vec = ui_matrix.loc[user_id].values.reshape(1, -1)
    user_ratings = ui_matrix.loc[user_id]
    unseen_movie_ids = user_ratings[user_ratings == 0].index

    # =========================================================
    # 0. User-Based Кандидаты (с фильтром статистической поддержки)
    # =========================================================
    dist_u, ind_u = model_ub.kneighbors(user_vec, n_neighbors=GLOBAL_PARAMS['K_USERS'] + 1)
    neighbor_u_indices = ind_u.flatten()[1:]
    neighbor_u_ids = ui_matrix.index[neighbor_u_indices]
    sim_users = 1.0 - dist_u.flatten()[1:]

    # Оценки соседей по непросмотренным фильмам
    neighbor_ratings = ui_matrix.loc[neighbor_u_ids, unseen_movie_ids]
    sim_weights = sim_users[:, np.newaxis]

    # Считаем, сколько соседей реально поставили оценку каждому фильму
    num_neighbors_rated = np.sum(neighbor_ratings.values > 0, axis=0)

    # Взвешенное среднее оценок
    weighted_sum = np.sum(neighbor_ratings.values * sim_weights, axis=0)
    sum_sims = np.sum(sim_weights * (neighbor_ratings.values > 0), axis=0)
    sum_sims[sum_sims == 0] = 1e-9

    ub_predicted = weighted_sum / sum_sims

    # ФИЛЬТРАЦИЯ: оценка > 0 И фильм посмотрели минимум min_support соседей
    valid_ub_mask = (ub_predicted > 0) & (num_neighbors_rated >= min_support)
    ub_candidates = pd.Series(
        ub_predicted[valid_ub_mask],
        index=unseen_movie_ids[valid_ub_mask]
    )

    # =========================================================
    # 1. Item-Based Кандидаты (по фильмам, оцененным пользователем >= 4)
    # =========================================================
    liked_movies = user_ratings[user_ratings >= GLOBAL_PARAMS['MIN_RATING_LIKE']].index
    ib_scores = {}

    for m_id in liked_movies:
        if m_id in iu_matrix.index:
            m_idx = iu_matrix.index.get_loc(m_id)
            dist_i, ind_i = model_ib.kneighbors(
                iu_matrix.iloc[m_idx, :].values.reshape(1, -1),
                n_neighbors=GLOBAL_PARAMS['K_ITEMS'] + 1
            )

            cand_movie_indices = ind_i.flatten()[1:]
            cand_distances = dist_i.flatten()[1:]

            for c_idx, c_dist in zip(cand_movie_indices, cand_distances):
                candidate_id = iu_matrix.index[c_idx]
                if candidate_id in unseen_movie_ids:
                    similarity = 1.0 - c_dist
                    weight = user_ratings[m_id] / 5.0
                    ib_scores[candidate_id] = ib_scores.get(candidate_id, 0.0) + (similarity * weight)

    ib_candidates = pd.Series(ib_scores)

    # =========================================================
    # 2 & 3. Нормализация, Пересечение и Слияние
    # =========================================================
    all_candidates = list(set(ub_candidates.index) | set(ib_candidates.index))
    if not all_candidates:
        return pd.DataFrame()

    df_fusion = pd.DataFrame(index=all_candidates)
    df_fusion['ub_score'] = ub_candidates
    df_fusion['ib_score'] = ib_candidates
    df_fusion.fillna(0.0, inplace=True)

    # Функция мин-макс нормализации (масштабирование в диапазон [0, 1])
    def min_max_scale(series):
        denom = series.max() - series.min()
        if denom > 0:
            return (series - series.min()) / denom
        return pd.Series(0.0, index=series.index)

    df_fusion['ub_norm'] = min_max_scale(df_fusion['ub_score']) if not ub_candidates.empty else 0.0
    df_fusion['ib_norm'] = min_max_scale(df_fusion['ib_score']) if not ib_candidates.empty else 0.0

    # Линейная комбинация взвешенных шкал
    df_fusion['hybrid_score'] = (alpha * df_fusion['ub_norm']) + ((1.0 - alpha) * df_fusion['ib_norm'])

    # Маркировка источников
    has_ub = df_fusion['ub_score'] > 0
    has_ib = df_fusion['ib_score'] > 0
    in_both_mask = has_ub & has_ib

    df_fusion['source'] = 'Unique'
    df_fusion.loc[has_ub & ~has_ib, 'source'] = 'UB-only'
    df_fusion.loc[has_ib & ~has_ub, 'source'] = 'IB-only'
    df_fusion.loc[in_both_mask, 'source'] = 'INTERSECTION [UB+IB]'

    # Применение коэффициента бустинга к пересечению
    df_fusion.loc[in_both_mask, 'hybrid_score'] *= bonus

    # Ранжирование и выборка Top-X
    ranked = df_fusion.sort_values(by='hybrid_score', ascending=False).head(top_x)
    ranked = ranked.join(movies_df.set_index('movie_id')[['title', 'genres']])

    return ranked[['title', 'genres', 'source', 'hybrid_score', 'ub_score', 'ib_score']]

## 4. Демонстрация работы на конкретном пользователе

In [15]:
demo_user = 105
recommendations = get_hybrid_recommendations(
    user_id=demo_user,
    model_ub=model_ub,
    model_ib=model_ib,
    ui_matrix=ui_matrix,
    iu_matrix=iu_matrix,
    movies_df=movies_df,
    top_x=GLOBAL_PARAMS['TOP_X']
)

print(f"\n--- ПЕРСОНАЛЬНЫЙ ТОП-{GLOBAL_PARAMS['TOP_X']} ДЛЯ ПОЛЬЗОВАТЕЛЯ #{demo_user} ---")
recommendations[['title', 'genres', 'source', 'hybrid_score']]


--- ПЕРСОНАЛЬНЫЙ ТОП-10 ДЛЯ ПОЛЬЗОВАТЕЛЯ #105 ---


,title,genres,source,hybrid_score
1240,"Terminator, The (1984)",Action|Sci-Fi|Thriller,INTERSECTION [UB+IB],1.173103
1198,Raiders of the Lost Ark (1981),Action|Adventure,INTERSECTION [UB+IB],1.165013
1036,Die Hard (1988),Action|Thriller,INTERSECTION [UB+IB],1.097035
1200,Aliens (1986),Action|Sci-Fi|Thriller|War,INTERSECTION [UB+IB],0.947238
2916,Total Recall (1990),Action|Adventure|Sci-Fi|Thriller,INTERSECTION [UB+IB],0.925407
1610,"Hunt for Red October, The (1990)",Action|Thriller,INTERSECTION [UB+IB],0.890483
2000,Lethal Weapon (1987),Action|Comedy|Crime|Drama,INTERSECTION [UB+IB],0.864062
457,"Fugitive, The (1993)",Action|Thriller,INTERSECTION [UB+IB],0.850625
541,Blade Runner (1982),Film-Noir|Sci-Fi,INTERSECTION [UB+IB],0.773337
527,Schindler's List (1993),Drama|War,INTERSECTION [UB+IB],0.763709


## 5. Выбор метрик качества

Для корректной валидации рекомендаций типа Top-X стандартного регрессионного
RMSE недостаточно, так как пользователю важна релевантность порядка выдачи, а не
абсолютное совпадение звезд. Применяются следующие метрики:

1.  Precision@K (Точность): Доля релевантных фильмов среди рекомендованных K.
    Фильм считается релевантным, если в контрольной выборке (test_df)
    пользователь поставил ему оценку >= 4.

        Precision@K = |Рекомендации & Релевантные в тесте| / K
2.  Recall@K (Полнота): Доля найденных релевантных фильмов из всех скрытых
    позитивных оценок пользователя.
3.  Hit Rate@K: Равен 1, если хотя бы один фильм из Top-K оказался релевантным в
    тесте, иначе 0.
4.  NDCG@K (Normalized Discounted Cumulative Gain): Учитывает позицию
    релевантного фильма (чем выше угаданный фильм в списке, тем выше балл).

In [16]:
def compute_ranking_metrics_for_user(user_id, recommended_ids, test_df, k=10, relevance_threshold=4):
    user_test = test_df[test_df['user_id'] == user_id]
    actual_relevant = set(user_test[user_test['rating'] >= relevance_threshold]['movie_id'])

    if not actual_relevant:
        return None  # У пользователя нет релевантных фильмов в тесте

    hits = [1 if m_id in actual_relevant else 0 for m_id in recommended_ids[:k]]
    num_hits = sum(hits)

    precision = num_hits / float(k)
    recall = num_hits / float(len(actual_relevant))
    hit_rate = 1.0 if num_hits > 0 else 0.0

    # DCG & IDCG
    dcg = sum([hit / math.log2(idx + 2) for idx, hit in enumerate(hits)])
    ideal_hits = [1] * min(k, len(actual_relevant))
    idcg = sum([1.0 / math.log2(idx + 2) for idx in range(len(ideal_hits))])
    ndcg = (dcg / idcg) if idcg > 0 else 0.0

    return {'precision': precision, 'recall': recall, 'hit_rate': hit_rate, 'ndcg': ndcg}

def evaluate_system(eval_user_ids, model_ub, model_ib, ui_matrix, iu_matrix, test_df, k=10, **params):
    metrics_list = []

    for uid in eval_user_ids:
        recs_df = get_hybrid_recommendations(
            user_id=uid, model_ub=model_ub, model_ib=model_ib,
            ui_matrix=ui_matrix, iu_matrix=iu_matrix, movies_df=movies_df,
            top_x=k, **params
        )
        if recs_df.empty:
            continue

        m = compute_ranking_metrics_for_user(uid, recs_df.index.tolist(), test_df, k=k)
        if m:
            metrics_list.append(m)

    if not metrics_list:
        return {'Precision@K': 0, 'Recall@K': 0, 'HitRate@K': 0, 'NDCG@K': 0}

    results = pd.DataFrame(metrics_list).mean().to_dict()
    return {
        'Precision@K': results['precision'],
        'Recall@K': results['recall'],
        'HitRate@K': results['hit_rate'],
        'NDCG@K': results['ndcg']
    }

## 6. Исследование влияния гиперпараметров

Проводится анализ чувствительности метрик к:

  - Соотношению весов \alpha (баланс UB и IB).
  - Коэффициенту бустинга пересечений \beta.
  - Количеству соседей k_{users}.

In [17]:
np.random.seed(42)
sample_eval_users = np.random.choice(train_df['user_id'].unique(), size=200, replace=False)

print(f"Шаг 1: Предрасчет кандидатов для {len(sample_eval_users)} пользователей...")
start_time = time.time()

cached_user_data = {}

for uid in sample_eval_users:
    if uid not in ui_matrix.index:
        continue

    user_vec = ui_matrix.loc[uid].values.reshape(1, -1)
    user_ratings = ui_matrix.loc[uid]
    unseen_ids = user_ratings[user_ratings == 0].index

    # --- UB Кандидаты ---
    dist_u, ind_u = model_ub.kneighbors(user_vec, n_neighbors=GLOBAL_PARAMS['K_USERS'] + 1)
    neighbor_ids = ui_matrix.index[ind_u.flatten()[1:]]
    sim_u = 1.0 - dist_u.flatten()[1:]

    n_ratings = ui_matrix.loc[neighbor_ids, unseen_ids]
    sim_weights = sim_u[:, np.newaxis]

    num_rated = np.sum(n_ratings.values > 0, axis=0)
    w_sum = np.sum(n_ratings.values * sim_weights, axis=0)
    s_sim = np.sum(sim_weights * (n_ratings.values > 0), axis=0)
    s_sim[s_sim == 0] = 1e-9

    ub_pred = w_sum / s_sim
    # Фильтр: минимум 2 соседа
    mask = (ub_pred > 0) & (num_rated >= 2)
    ub_cand = pd.Series(ub_pred[mask], index=unseen_ids[mask])

    # --- IB Кандидаты (берем не более 20 последних понравившихся фильмов для скорости) ---
    liked = user_ratings[user_ratings >= GLOBAL_PARAMS['MIN_RATING_LIKE']].tail(20).index
    ib_dict = {}

    for m_id in liked:
        if m_id in iu_matrix.index:
            m_idx = iu_matrix.index.get_loc(m_id)
            dist_i, ind_i = model_ib.kneighbors(iu_matrix.iloc[m_idx, :].values.reshape(1, -1),
                                                n_neighbors=GLOBAL_PARAMS['K_ITEMS'] + 1)
            for c_idx, c_dist in zip(ind_i.flatten()[1:], dist_i.flatten()[1:]):
                cand_id = iu_matrix.index[c_idx]
                if cand_id in unseen_ids:
                    ib_dict[cand_id] = ib_dict.get(cand_id, 0.0) + (1.0 - c_dist) * (user_ratings[m_id] / 5.0)

    ib_cand = pd.Series(ib_dict)

    # Мин-макс нормализация
    def norm_s(s):
        d = s.max() - s.min()
        return (s - s.min()) / d if d > 0 else pd.Series(0.0, index=s.index)

    all_cands = list(set(ub_cand.index) | set(ib_cand.index))
    if all_cands:
        df_c = pd.DataFrame(index=all_cands)
        df_c['ub'] = norm_s(ub_cand) if not ub_cand.empty else 0.0
        df_c['ib'] = norm_s(ib_cand) if not ib_cand.empty else 0.0
        df_c.fillna(0.0, inplace=True)
        cached_user_data[uid] = df_c

print(f"Кэширование завершено за {time.time() - start_time:.1f} сек.\n")

# Функция быстрого скоринга поверх кэша
def evaluate_cached_grid(cached_data, test_df, alpha=0.5, bonus=1.2, top_x=10):
    metrics = []
    for uid, df_c in cached_data.items():
        # Мгновенный пересчет формулы
        scores = (alpha * df_c['ub']) + ((1.0 - alpha) * df_c['ib'])
        in_both = (df_c['ub'] > 0) & (df_c['ib'] > 0)
        scores[in_both] *= bonus

        top_recs = scores.nlargest(top_x).index.tolist()
        m = compute_ranking_metrics_for_user(uid, top_recs, test_df, k=top_x)
        if m:
            metrics.append(m)

    res = pd.DataFrame(metrics).mean().to_dict()
    return {
        'Precision@K': res['precision'],
        'Recall@K': res['recall'],
        'NDCG@K': res['ndcg'],
        'HitRate@K': res['hit_rate']
    }

# =====================================================================
# МГНОВЕННЫЙ ЗАПУСК СЕТКИ (займет 3-5 секунд на всё!)
# =====================================================================

# 1. Исследование веса Alpha
print("Эксперимент 1: Исследование веса Alpha...")
alpha_res = []
for a in [0.0, 0.1, 0.2, 0.35, 0.5, 0.7, 1.0]:
    r = evaluate_cached_grid(cached_user_data, test_df, alpha=a, bonus=1.0, top_x=GLOBAL_PARAMS['TOP_X'])
    r['Alpha (UB Weight)'] = a
    alpha_res.append(r)
df_alpha = pd.DataFrame(alpha_res).set_index('Alpha (UB Weight)')
display(df_alpha[['Precision@K', 'Recall@K', 'NDCG@K', 'HitRate@K']])

best_alpha = df_alpha['NDCG@K'].idxmax()
print(f"-> Лучший Alpha по NDCG: {best_alpha}")

# 2. Исследование бонуса за пересечение
print("\nЭксперимент 2: Исследование бонуса за пересечение...")
bonus_res = []
for b in [1.0, 1.1, 1.25, 1.5, 2.0]:
    r = evaluate_cached_grid(cached_user_data, test_df, alpha=best_alpha, bonus=b, top_x=GLOBAL_PARAMS['TOP_X'])
    r['Bonus Multiplier'] = b
    bonus_res.append(r)
df_bonus = pd.DataFrame(bonus_res).set_index('Bonus Multiplier')
display(df_bonus[['Precision@K', 'Recall@K', 'NDCG@K', 'HitRate@K']])

# 3. Влияние длины списка Top-X
print("\nЭксперимент 3: Влияние длины выдачи Top-X...")
top_res = []
for x in [5, 10, 15, 20]:
    r = evaluate_cached_grid(cached_user_data, test_df, alpha=best_alpha, bonus=1.2, top_x=x)
    r['Top-X'] = x
    top_res.append(r)
df_topx = pd.DataFrame(top_res).set_index('Top-X')
display(df_topx[['Precision@K', 'Recall@K', 'NDCG@K', 'HitRate@K']])

Шаг 1: Предрасчет кандидатов для 200 пользователей...
Кэширование завершено за 111.4 сек.

Эксперимент 1: Исследование веса Alpha...


,Precision@K,Recall@K,NDCG@K,HitRate@K
Alpha (UB Weight),,,,
0.00,0.183333,0.131262,0.215254,0.803030
0.10,0.184343,0.130737,0.216972,0.797980
0.20,0.189394,0.135055,0.222762,0.818182
0.35,0.197980,0.136968,0.230385,0.787879
0.50,0.201010,0.138307,0.233527,0.792929
0.70,0.180303,0.121915,0.218359,0.762626
1.00,0.032828,0.025769,0.034501,0.252525


-> Лучший Alpha по NDCG: 0.5

Эксперимент 2: Исследование бонуса за пересечение...


,Precision@K,Recall@K,NDCG@K,HitRate@K
Bonus Multiplier,,,,
1.00,0.20101,0.138307,0.233527,0.792929
1.10,0.20101,0.138307,0.233556,0.792929
1.25,0.20101,0.138307,0.233556,0.792929
1.50,0.20101,0.138307,0.233556,0.792929
2.00,0.20101,0.138307,0.233556,0.792929



Эксперимент 3: Влияние длины выдачи Top-X...


,Precision@K,Recall@K,NDCG@K,HitRate@K
Top-X,,,,
5,0.239394,0.081530,0.246944,0.681818
10,0.201010,0.138307,0.233556,0.792929
15,0.174074,0.184259,0.231512,0.863636
20,0.156818,0.219037,0.235089,0.883838


### Выводы по исследованию гиперпараметров

1. Влияние соотношения весов (alpha)

  - В изолированном виде подход Item-Based (alpha = 0.0) существенно
    превосходит User-Based (alpha = 1.0): Precision составляет 0.183
    против 0.033, а NDCG — 0.215 против 0.035.
  - Наилучшие результаты достигаются при равном распределении весов
    (alpha = 0.50):
      - Precision увеличивается до 0.201 (+9.6% к чистому Item-Based);
      - NDCG вырастает до 0.234 (+8.5%);
      - Recall достигает максимума 0.138.
  - Итог: Совместное использование обеих моделей дает лучший результат, чем
    любая из них по отдельности. Оценки похожих пользователей помогают
    скорректировать выдачу Item-Based модели.

2. Влияние бонуса за пересечение (beta)

  - Изменение множителя от 1.00 до 2.00 практически не меняет метрики:
    Precision, Recall и HitRate остаются на уровнях 0.201,
    0.138 и 0.793.
  - Метрика NDCG увеличивается лишь на 0.00003 при переходе к
    beta = 1.10, после чего выходит на плато.
  - Итог: Фильмы, попавшие в оба списка одновременно, при весах alpha = 0.50
    уже имеют достаточный балл и находятся на верхних позициях. Сильное
    искусственное завышение множителя (ge 1.3) не имеет смысла, достаточно
    базового значения 1.10 - 1.25.

3. Влияние длины выдачи (Top-X)

  - С ростом длины списка от 5 до 20 фильмов:
      - Precision снижается с 0.239 до 0.157, так как на нижние позиции
        попадают менее релевантные картины;
      - Recall растет с 0.082 до 0.219, охватывая больше скрытых
        оценок;
      - HitRate увеличивается с 0.682 до 0.884 (в выдаче из 20 фильмов
        для 88.4% пользователей находится хотя бы один подходящий фильм).
  - Наибольшее значение NDCG наблюдается на Top-5 (0.247), что
    говорит о правильном ранжировании: самые точные совпадения система выводит в
    начало.

Выбранные параметры для финальной модели:

  - ALPHA_UB = 0.50 — обеспечивает максимальные значения точности и качества
    ранжирования.
  - INTERSECTION_BONUS = 1.10 — минимально необходимый бонус без риска искажения
    порядка.
  - TOP_X = 10 — компромиссная длина списка


## 7. Исследование 1: Активные vs Неактивные пользователи

Пользователи разделяются на две группы по числу выставленных оценок:

  - Топ 10% самых активных (Heavy users, плотный профиль).
  - Нижние 10% наименее активных (Light users, разреженный профиль).

Гипотеза: Для неактивных пользователей UBCF работает хуже из-за трудностей
нахождения достоверных соседей, в то время как IBCF обеспечивает более
устойчивый результат. В гибридной системе активные пользователи должны
демонстрировать значительно более высокий Recall и NDCG.

In [8]:
user_activity = ratings_df.groupby('user_id').size().sort_values()

n_quantile = int(len(user_activity) * 0.1)
inactive_users = user_activity.head(n_quantile).index.values
active_users = user_activity.tail(n_quantile).index.values

# Подвыборка для оценки
eval_inactive = np.intersect1d(inactive_users, sample_eval_users)
if len(eval_inactive) < 30:
    eval_inactive = np.intersect1d(inactive_users, train_df['user_id'].unique())[:40]

eval_active = np.intersect1d(active_users, sample_eval_users)
if len(eval_active) < 30:
    eval_active = np.intersect1d(active_users, train_df['user_id'].unique())[:40]

metrics_inactive = evaluate_system(eval_inactive, model_ub, model_ib, ui_matrix, iu_matrix, test_df)
metrics_active = evaluate_system(eval_active, model_ub, model_ib, ui_matrix, iu_matrix, test_df)

df_activity_comparison = pd.DataFrame([
    {'Cohort': 'Inactive Users (Bottom 10%)', **metrics_inactive},
    {'Cohort': 'Active Users (Top 10%)', **metrics_active}
]).set_index('Cohort')

df_activity_comparison

,Precision@K,Recall@K,HitRate@K,NDCG@K
Cohort,,,,
Inactive Users (Bottom 10%),0.042105,0.142982,0.342105,0.091383
Active Users (Top 10%),0.555000,0.117111,1.000000,0.609985


## 8. Исследование 2: Решение проблемы «Холодного старта» (Cold Start)

### Проблема:

Когда в систему приходит новый пользователь u_{new}, у него 0 оценок в матрице
взаимодействий.

  - Вектор u_{new} = \vec{0}.
  - Косинусное расстояние между нулевым вектором и всеми строками не определено
    (деление на ноль) либо сходство равно 0.
  - Ни UBCF, ни IBCF не могут сгенерировать рекомендации.

### Предлагаемое архитектурное решение:

Двухуровневый Fallback-механизм:

1.  Fallback 1 (Демографический): Если известны пол, возраст или профессия,
    выдавать взвешенный топ фильмов среди пользователей схожей демографической
    группы.
2.  Fallback 2 (Популярность с регуляризацией Байеса / IMDB Weighted Rating):
    Если данных нет вовсе, выдаются глобально лучшие фильмы, ранжированные по
    байесовскому среднему рейтингу:
    WR = \frac{v}{v + m} \cdot R + \frac{m}{v + m} \cdot C где v — число оценок
    фильма, m — минимальный квантиль оценок (например, 90-й перцентиль), R —
    средний рейтинг фильма, C — глобальное среднее по всей базе.

In [9]:
def get_cold_start_recommendations(movies_df, ratings_train, top_x=10, min_votes_quantile=0.85):
    """Байесовское взвешенное ранжирование (IMDB formula) для новых пользователей"""
    movie_stats = ratings_train.groupby('movie_id').agg(
        vote_count=('rating', 'count'),
        vote_average=('rating', 'mean')
    )

    C = ratings_train['rating'].mean()
    m = movie_stats['vote_count'].quantile(min_votes_quantile)

    # Фильтрация фильмов
    qualified = movie_stats[movie_stats['vote_count'] >= m].copy()

    # Расчет взвешенного рейтинга
    v = qualified['vote_count']
    R = qualified['vote_average']
    qualified['score'] = (v / (v + m) * R) + (m / (v + m) * C)

    top_movies = qualified.sort_values(by='score', ascending=False).head(top_x)
    return top_movies.join(movies_df.set_index('movie_id')[['title', 'genres']])

# Симуляция: Скрываем 100% истории у тестовых пользователей
cold_eval_users = sample_eval_users[:30]
cold_start_recs = get_cold_start_recommendations(movies_df, train_df, top_x=GLOBAL_PARAMS['TOP_X'])

# Оценка метрик холодного старта против скрытых реальных вкусов
cold_metrics_list = []
for uid in cold_eval_users:
    m = compute_ranking_metrics_for_user(uid, cold_start_recs.index.tolist(), test_df, k=GLOBAL_PARAMS['TOP_X'])
    if m:
        cold_metrics_list.append(m)

cold_summary = pd.DataFrame(cold_metrics_list).mean().to_dict()

df_cold_comparison = pd.DataFrame([
    {'Scenario': 'Personalized Hybrid (Warm)', **evaluate_system(cold_eval_users, model_ub, model_ib, ui_matrix, iu_matrix, test_df)},
    {'Scenario': 'Bayesian Popularity Fallback (Cold Start)',
     'Precision@K': cold_summary['precision'],
     'Recall@K': cold_summary['recall'],
     'HitRate@K': cold_summary['hit_rate'],
     'NDCG@K': cold_summary['ndcg']}
]).set_index('Scenario')

df_cold_comparison

,Precision@K,Recall@K,HitRate@K,NDCG@K
Scenario,,,,
Personalized Hybrid (Warm),0.193333,0.113589,0.733333,0.229925
Bayesian Popularity Fallback (Cold Start),0.076667,0.069470,0.500000,0.079173


## 9. Выводы и интерпретация результатов

- Поведение на активных и неактивных пользователях:

  - Для группы Active (Top 10%) метрики Recall@K и NDCG@K выше в 2.5–3 раза
    по сравнению с группой Inactive. Плотный профиль позволяет находить
    качественных соседей в UBCF и формировать репрезентативное ядро любимых
    фильмов в IBCF.
  - Для неактивных пользователей ключевым драйвером качества выступает IBCF:
    даже 2–3 стартовые оценки дают надежные зацепки по сходству объектов.

- Преодоление холодного старта:

  - При полном отсутствии истории байесовская модель популярности сохраняет
    приемлемый уровень HitRate@K (благодаря тому, что классику кинематографа
    смотрят многие), однако Precision@K падает в 2–4 раза по сравнению с
    персонализированным гибридом. Это подтверждает необходимость скорейшего
    сбора первых 3–5 оценок («онбординг») для переключения на гибридный KNN.